data-analytics/dataflow-python-examples/dataflow_python_examples/data_generation_for_benchmarking.py

```diff
@@ -0,0 +1,485 @@
+                                                      self.data_gen.max_int)
+            if self.data_gen.only_pos and record[fieldname] < 0:
+                record[fieldname] = abs(record[fieldname])
+            record[fieldname] = int(record[fieldname])
+        elif field[u'type'] == 'FLOAT':
+            # This implements max and sign constraints
+            record[fieldname] = faker.pyfloat(math.log10(self.data_gen.max_float),
+                                              self.data_gen.float_precision,
+                                              self.data_gen.only_pos)
+            record[fieldname] = float(record[fieldname])
+
+        # Make some values null based on null_prob.
+        if field[u'mode'] == 'NULLABLE':
+            record[fieldname] = np.random.choice([None, record[fieldname]],
+                                                 p=[self.data_gen.null_prob,
+                                                    1.0 - self.data_gen.null_prob])
+
+        # Pick key at random from foreign keys.
+        # Draw key column from [0, n_keys) if has _key in the name.
+        # This forces key column to no contain nulls
+        if ('_key' in field[u'name'].lower()) or ('_id' in field[u'name'].lower()):
+            key_mag = int(math.log10(self.data_gen.n_keys))
+            # Assume the key field is of string type and format it to be left zero padded.
+            record[fieldname] = format(np.random.randint(0, self.data_gen.n_keys),
+                                       '0' + str(key_mag + 1) + 'd')
+        return record
+
+    def generate_fake(self, fschema):
+        """
+        This method creates a single fake record based on the constraints defined int
+        the FakeRowGen instance's data_gen attribute.
+
+        Arguments:
+                fschema: A dictionary containing a faker_schema (this should be generated by
+                         DataGenerator.get_faker_schema() )
+        """
+        # Initialize a FakerSchema object.
+        schema_faker = FakerSchema()
+
+        # Generate a fake record.
+        data = schema_faker.generate_fake(fschema, 1)  # Generate one record.
+
+        # This performs a sanity check on datatypes and parameterized constraints.
+        for col_name in data:
+            data = self.sanity_check(data, col_name)
+
+        return data
+
+    def process(self, element, *args, **kwargs):
+        """This function creates n random data records based on the properties
+        of the passed DataGenerator object for each element in prior the PCollection.
+
+        Args:
+            element: A single element of the PCollection (the contents of this element are ignored
+                     in this particularDoFn so we just pass it newlines).
+        """
+
+        faker_schema = self.data_gen.get_faker_schema()
+
+        row = [self.generate_fake(faker_schema)]
+        return row  # PCollection returned by DoFn.process() must be an iterable.
+
+
+def run(argv=None):
```

Comment:
This function is way too long for a single function. I'd suggest breaking it up.

Reply:
Teasing out logic into separate functions. run() will be reduced to:
```python
def run(argv=None):
    """
    This funciton parses the command line arguments and runs the Beam Pipeline.

    Args:
        argv: list containing the commandline arguments for this call of the script.
    """
    schema_inferred = False

    data_args, pipeline_args = parse_data_generator_args(argv)
    data_args, schema_inferred = fetch_schema(data_args, schema_inferred)
    pipeline_options = PipelineOptions(pipeline_args)

    temp_location = pipeline_options.display_data()['temp_location']
    temp_blob = write_n_line_file_to_gcs(temp_location, data_args.num_records)

    data_gen = data_generator_from_data_args(data_args)
    # Initiate the pipeline using the pipeline arguments passed in from the
    # command line.  This includes information including where Dataflow should
    # store temp files, and what the project id is and what runner to use.
    p = beam.Pipeline(options=pipeline_options)

    (p

     # Read the file we created with num_records newlines.

     | 'Read file with num_records lines' >> beam.io.ReadFromText(temp_location +
                                                                  '/temp_num_records.txt')
     # Use our instance of our custom DataGenerator Class to generate 1 fake datum
     # with the appropriate schema for each element in the PColleciton created above.
     | 'Generate Data' >> beam.ParDo(FakeRowGen(data_gen))
     | 'Write to BigQuery' >> beam.io.gcp.bigquery.WriteToBigQuery(
         # The table name is a required argument for the BigQuery sink.
         # In this case we use the value passed in from the command line.
         data_args.output_bq_table,
         schema=None if schema_inferred else data_gen.get_bq_schema_string(),
         # Creates the table in BigQuery if it does not yet exist.
         create_disposition=beam.io.BigQueryDisposition.CREATE_IF_NEEDED,
         write_disposition=data_gen.write_disp,
         # Use the max recommended batch size.
         batch_size=500)
    )

    p.run().wait_until_finish()

    # Manually clean up of temp_num_records.txt because it will be outside this job's
    # directory and Dataflow will not remove it for us.
    temp_blob.delete()

```